imxrt-hal/src/adc.rs

```diff
@@ -0,0 +1,275 @@
+//!
+//! let mut adc1 = adc1_builder.build(adc::ClockSelect::default(), adc::ClockDivision::default());
+//! let mut a1 = adc::AnalogInput::new(peripherals.iomuxc.ad_b1.p02, &adc1);
+//!
+//! let reading: u16 = adc1.read(&mut a1).unwrap();
+//!```
+
+use crate::ccm;
+use crate::iomuxc::adc::{prepare_adc_pin, Pin, ADC1, ADC2};
+use crate::iomuxc::{adc, consts::Unsigned};
+use crate::ral;
+use core::marker::PhantomData;
+use embedded_hal::adc::{Channel, OneShot};
+
+/// The clock input for an ADC
+#[allow(non_camel_case_types)]
+pub enum ClockSelect {
+    IPG,   // IPG clock
+    IPG_2, // IPG clock / 2
+    ADACK, // Asynchronous clock
+}
+
+/// How much to divide the clock input
+pub enum ClockDivision {
+    Div1, // Input clock / 1
+    Div2, // Input clock / 2
+    Div4, // Input clock / 4
+    Div8, // Input clock / 8
+}
+
+impl ClockSelect {
+    pub fn default() -> Self {
+        ClockSelect::ADACK
+    }
+}
+
+impl ClockDivision {
+    pub fn default() -> Self {
+        ClockDivision::Div2
+    }
+}
+
+/// Conversion speeds done by clock cycles
+pub enum ConversionSpeed {
+    Slow,     // 25 ADC clock cycles (24 on imxrt102x)
+    Medium,   // 17 ADC clock cycles (16 on imxrt102x)
+    Fast,     // 9 ADC clock cycles (8 on imxrt102x)
+    VeryFast, // 3 ADC clock cycles (2 on imxrt102x)
+}
+
+/// Denotes how much hardware averaging to do
+pub enum AveragingCount {
+    Avg1,
+    Avg4,
+    Avg8,
+    Avg16,
+    Avg32,
+}
+
+// Specifies the resolution the ADC
+pub enum ResolutionBits {
+    Res8,
+    Res10,
+    Res12,
+}
+
+/// A pin representing an analog input for a particular ADC
+pub struct AnalogInput<ADCx, P> {
+    _module: PhantomData<ADCx>,
+    _pin: P,
+}
+
+impl<P, ADCx> Channel<ADCx> for AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    type ID = u16;
+
+    fn channel() -> Self::ID {
+        <P as Pin<ADCx>>::Input::U16
+    }
+}
+
+impl<P, ADCx> AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    /// Creates a new analog input pin
+    pub fn new(mut pin: P, _adc: &ADC<ADCx>) -> Self {
+        prepare_adc_pin(&mut pin);
+        Self {
+            _module: PhantomData,
+            _pin: pin,
+        }
+    }
+}
+
+pub struct ADC<ADCx> {
+    _module: PhantomData<ADCx>,
+    reg: ral::adc::Instance,
+}
+
+impl<ADCx> ADC<ADCx> {
+    fn new(reg: ral::adc::Instance) -> Self {
+        let inst = Self {
+            _module: PhantomData,
+            reg,
+        };
+
+        inst.set_resolution(ResolutionBits::Res10);
+        inst.set_low_power_mode(false);
+
+        // Calibrate w/ slow settings initially
+        inst.set_averaging(AveragingCount::Avg32);
+        inst.set_conversion_speed(ConversionSpeed::Slow);
+        inst.calibrate();
+
+        // Set to default of 4 hardware averages & medium conversion speed
+        inst.set_averaging(AveragingCount::Avg4);
+        inst.set_conversion_speed(ConversionSpeed::Medium);
+
+        inst
```

Comment:
I think the configuration functions above make sense, I do have a concern about them once interrupts and some sort of streaming comes about. What happens if you tell the ADC to stream conversions and then change the settings while its streaming? Would it make more sense to provide a Configuration struct and let the ADC then configure itself as needed entirely? Would it save any register modify calls? Would it make changing the configuration in an atomic (from the callers perspective) easier? Something to maybe consider

Reply:
Fair point. While it wouldn't save any register modify calls, I'm fine with making a configuration struct. I'm not sure what you mean about streaming conversion, though. Do you mean putting an ADC in a continuous mode? I decided not to do continuous mode for the time being because `embedded-hal` has no traits for it. But in that case I feel like disabling any configuration changes while streaming is the easiest way of going about it, however that be implemented. Otherwise, for the one shot reads, I think putting them in an interrupt free block would be the easiest way there (assuming there aren't any implications with that, I'm not familiar with that side of the code).

Reply:
As long as the only method for using the ADC is a polling method where the code flow is blocked, this is fine and safe and friendly.  In an interrupt driven conversion you really want to change the config of the ADC in a somewhat atomic way, either stopping the ADC, changing config, starting it again, or dropping values you know don't care about while the config is changing. You don't want your desired ADC config to be up in the air as to what it actually is. That can be done by using a configure(cfg: AdcConfiguration) type function for people, or it can be done by users themselves however they like. It's something to keep in mind. I don't think one way or the other is bad, the functions you have are needed regardless and we can add on an atomic-like configure function later.